### PR TITLE
[IN-154][OSF] Add collectedGuidMetadata to OSF Elasticsearch Index

### PR DESCRIPTION
**Pull request by fabmiz**
## Purpose

Index CollectedGuidMetadata in OSF Elasticsearch

## Changes

- Add async methods to update ES index 
- Update docs on privacy changes
- Add tests

## Side Effects

N/A

## Ticket

[IN-154](https://openscience.atlassian.net/browse/IN-154)

### CODE REVIEW

File: website/search/elastic_search.py [hunk 1]
```diff
@@ -580,6 +626,38 @@ def update_institution(institution, index=None):
...
+                try:
+                    update_cgm(cgm, op=op, index=index)
+                except Exception as exc:
+                    self.retry(exc=exc)
+    except CollectedGuidMetadata.DoesNotExist:
+        logger.exception('Could not find object with <_id {}> in collection'.format(cgm_id))
+        return
+
+@requires_search
+def update_cgm(cgm, op='update', index=None):
```
[the remark this answers is not in the document]
Comment (adlius) [reply website/search/elastic_search.py]: And I guess another reasoning behind the suggestion was to handle the logic of deleting index inside the `update_cgm` method. For example, when a collection gets updated, it calls `update_cgm`, inside which will determine whether to remove the index or updating the index depending on the current state of `is_deleted`/`is_collected` or other similar flags.

File: osf_tests/test_elastic_search.py [hunk 2]
```diff
@@ -62,6 +71,176 @@ def wrapped(*args, **kwargs):
         return wrapped
     return test_wrapper
 
+class TestCollectionsSearch(OsfTestCase):
```
Comment (mfraezz) [inline osf_tests/test_elastic_search.py]: A lot of these tests are mostly the same, but do the opposite of each other. Consider collating them in such a way as to minimize set-up time without testing any less functionality. (e.g. combining `test_submissions_turned_private_are_deleted_from_index` and `test_submissions_turned_public_are_added_to_index` to go from public -> private -> public again, checking ES in between, etc.)

File: website/search_migration/migrate.py [hunk 3]
```diff
@@ -130,6 +130,13 @@ def migrate_users(index, delete, increment=10000):
             es_args={'raise_on_error': False})  # ignore 404s
         logger.info('{} users marked deleted'.format(total_users))
 
+def migrate_collected_guid_metadata(index, delete):
+    cgms = CollectedGuidMetadata.objects.all()
+    if delete:
```
[the remark this answers is not in the document]
Comment (mfraezz) [reply website/search_migration/migrate.py]: Update: there appear to be exactly two cases where `delete` is important -
* if the first branch of the conditional in `set_up_index` is hit -- when the alias is missing or malformed (seemingly should only happen if migrating from nothing, so probably doesn't matter)
* If an old  run of `migrate_search` has failed, leaving the next version of the index partially populated but orphaned (likely). This can be obviated by dropping anything from the "new" index before populating it.

File: osf/models/collection.py [hunk 4]
```diff
@@ -31,11 +37,43 @@ class Meta:
...
         return self.guid._id
 
+    @cached_property
+    def _es_doc_id(self):
+        return '{}-{}'.format(self._id, self.collection._id)
+
+    @classmethod
+    def load(cls, data, select_for_update=False):
+        if isinstance(data, basestring):
+            cgm_id, collection_id = data.split('-')
```
Comment (mfraezz) [inline osf/models/collection.py]: This can cause a `ValueError` if there isn't exactly one `-`

File: osf/models/collection.py [hunk 5]
```diff
@@ -31,11 +37,43 @@ class Meta:
     def _id(self):
         return self.guid._id
 
+    @cached_property
+    def _es_doc_id(self):
+        return '{}-{}'.format(self._id, self.collection._id)
+
+    @classmethod
+    def load(cls, data, select_for_update=False):
+        if isinstance(data, basestring):
```
Comment (mfraezz) [inline osf/models/collection.py]: This should also handle the case where `data` is an `int`, and `get` based on `pk`, as the superclass does

File: osf/models/collection.py [hunk 6]
```diff
@@ -31,11 +37,43 @@ class Meta:
     def _id(self):
         return self.guid._id
 
+    @cached_property
+    def _es_doc_id(self):
```
Comment (mfraezz) [inline osf/models/collection.py]: Convention usually dictates that `_id` the same as what gets passed in to `load` -- you may want to update `_id`'s behavior to this instead.

File: website/search_migration/migrate.py [hunk 7]
```diff
@@ -130,6 +130,15 @@ def migrate_users(index, delete, increment=10000):
             es_args={'raise_on_error': False})  # ignore 404s
         logger.info('{} users marked deleted'.format(total_users))
 
+def migrate_collected_metadata(index, delete):
```
Comment (mfraezz) [inline website/search_migration/migrate.py]: This should be telling ES to delete everything with the CGM type before calling `bulk_update`

File: website/search/elastic_search.py [hunk 8]
```diff
@@ -578,6 +625,52 @@ def update_institution(institution, index=None):
...
+def update_cgm_async(self, cgm_id, collection_id=None, op='update', index=None):
+    CollectedGuidMetadata = apps.get_model('osf.CollectedGuidMetadata')
+    if collection_id:
+        try:
+            cgm = CollectedGuidMetadata.objects.filter(
+                guid___id=cgm_id,
+                collection_id=collection_id,
+                collection__provider__isnull=False,
+                collection__deleted__isnull=True,
+                collection__is_bookmark_collection=False).first()
```
Comment (mfraezz) [inline website/search/elastic_search.py]: `first` will return `None` if there aren't any -- the `except` here will never be hit. Likewise below

Comment (fabmiz) [reply website/search/elastic_search.py]: ah that was left off in debugging some failing tests where cgm was not found because the collection was non-provided.

File: website/search/elastic_search.py [hunk 9]
```diff
@@ -246,8 +248,11 @@ def format_results(results):
             result['parent_title'] = parent_info.get('title') if parent_info else None
         elif result.get('category') in {'project', 'component', 'registration', 'preprint'}:
             result = format_result(result, result.get('parent_id'))
+        elif result.get('category') == 'collectionSubmission':
+            pass
```
Comment (mfraezz) [inline website/search/elastic_search.py]: Should this `continue` instead? Otherwise I'd expect the raw result to be delivered to the v1 front-end, which I think isn't supposed to happen.

File: website/search/elastic_search.py [hunk 10]
```diff
@@ -419,6 +424,48 @@ def bulk_update_nodes(serialize, nodes, index=None):
...
+        contributors = obj._contributors.filter(contributor__visible=True).order_by('contributor___order').values('fullname', 'guids___id', 'is_active')
+
+    return {
+        'id': cgm._id,
+        'abstract': getattr(obj, 'description', ''),
+        'collectedType': getattr(cgm, 'collected_type'),
+        'contributors': [serialize_cgm_contributor(contrib) for contrib in contributors],
+        'status': cgm.status,
+        'subjects': list(cgm.subjects.values_list('text', flat=True)),
+        'title': getattr(obj, 'title'),
```
Comment (mfraezz) [inline website/search/elastic_search.py]: Minor: These will throw `AttributeError`s for objects that don't have a `title` or `url`. This matters little now, but will whenever we allow objects other than nodes to be collected.

File: website/search/elastic_search.py [hunk 11]
```diff
@@ -573,6 +620,52 @@ def update_institution(institution, index=None):
...
+            logger.exception('Could not find object <_id {}> in a collection <_id {}>'.format(cgm_id, collection_id))
+        else:
+            if cgm and hasattr(cgm.guid.referent, 'is_public') and cgm.guid.referent.is_public:
+                try:
+                    update_cgm(cgm, op=op, index=index)
+                except Exception as exc:
+                    self.retry(exc=exc)
+    else:
+        try:
+            cgms = CollectedGuidMetadata.objects.filter(
```
Comment (mfraezz) [inline website/search/elastic_search.py]: This will never throw a `DoesNotExist`, consider removing `try` block

File: api/base/filters.py [hunk 12]
```diff
@@ -430,6 +430,8 @@ def postprocess_query_param(self, key, field_name, operation):
                 else self.model_class.primary_identifier_name
             )
             operation['op'] = 'in'
+            if not isinstance(operation['value'], list):
```
Comment (sloria) [inline api/base/filters.py]: I'm not sure this belongs here. The parsing of `filter[id]=123,abc` happens in `bulk_get_values`. Perhaps that method should handle when single values are passed.

File: api/collections/serializers.py [hunk 13]
```diff
@@ -152,7 +152,7 @@ class Meta:
         'subjects',
         'status',
     ])
-    id = IDField(source='_id', read_only=True)
+    id = IDField(source='guid._id', read_only=True)
```
Comment (sloria) [inline api/collections/serializers.py]: Will this result in a `JOIN` for every serialized "Collected Meta"? Can you verify that this isn't causing a large performance penalty? 

The Django Debug Toolbar is useful for this.

File: osf/models/collection.py [hunk 14]
```diff
@@ -29,13 +35,51 @@ class Meta:
 
     @cached_property
     def _id(self):
-        return self.guid._id
+        return '{}-{}'.format(self.guid._id, self.collection._id)
+
+    @classmethod
+    def load(cls, data, select_for_update=False):
+        if isinstance(data, int):
```
Comment (sloria) [inline osf/models/collection.py]: Why is this necessary? Where do we load by pk?

File: osf/models/collection.py [hunk 15]
```diff
@@ -107,13 +159,26 @@ def save(self, *args, **kwargs):
...
+
         if first_save:
             # Set defaults for M2M
             self.collected_types = ContentType.objects.filter(app_label='osf', model__in=['abstractnode', 'collection'])
             # Set up initial permissions
             self.update_group_permissions()
             self.get_group('admin').user_set.add(self.creator)
+
+        elif 'is_public' in saved_fields:
+            cgms = list(self.collectedguidmetadata_set.all())
```
Comment (sloria) [inline osf/models/collection.py]: Minor: Is it necessary to cast to a list here?

File: osf/models/collection.py [hunk 16]
```diff
@@ -107,13 +159,26 @@ def save(self, *args, **kwargs):
...
+
+            if self.is_public:
+                # Add all collection submissions back to ES index
+                self.bulk_update_search(cgms)
```
Comment (sloria) [inline osf/models/collection.py]: This should probably happen in celery (queued to run after the request), like we do for nodes. Otherwise, failures to update ES will result in the whole transaction being rolled back 😱

File: api/collections/serializers.py (hunk 13 above)
Comment (fabmiz) [reply api/collections/serializers.py]: (# of search results, # of queries, time (ms))

with <guid._id>: 	(1, 20, 44.83), (2, 37, 69.53), (3, 54, 107), (4, 71, 143.33)
with <_id>:	(1, 19, 39.65), (2, 35, 60.75), (3, 51, 99.65), (4, 67, 134)

Seems with id=<guid._id> we are making 1 additional query per search result compared to same scenario with id=<_id>.

**Comment by mfraezz** [comment]
This is slated for the release after next, dropping the ticket in RTM and will merge this after next release.

**Comment by mfraezz** [comment]
Merged via 03eade942192e305269b407494849bdad95bf828, closing
